core/transform.ml

```diff
@@ -0,0 +1,195 @@
+module type INTERFACE = sig
+  type state
+  type 'a result
+
+  val name : string
+
+  val program : state ->
+                Sugartypes.program ->
+                Sugartypes.program result
+
+  val sentence : state ->
+                 Sugartypes.sentence ->
+                 Sugartypes.sentence result
+end
+
+module type UNTYPED = sig
+  type state = Context.t
+  type 'a result = Result of { program: 'a;
+                               state: state }
+
+
+  val return : state -> 'a -> 'a result
+  val context : state -> Context.t
+
+  module type S = sig
+    module Untyped: sig
+      include INTERFACE with type state := state and type 'a result := 'a result
+    end
+  end
+
+  module Make: sig
+    module Transformer(T : sig
+                 val name : string
+                 val obj : SugarTraversals.map
+               end): sig
+      include INTERFACE with type state := state and type 'a result := 'a result
+    end
+  end
+end
+
+module Untyped : UNTYPED = struct
+  type 'a transformer = 'a -> 'a
+
+  type state = Context.t
+  type 'a result = Result of { program: 'a;
```

Comment:
Any particular reason for this being variant type rather than just a record? It seems to work without the data constructor.

Reply:
In an earlier iteration it was relevant to allow multiple fields named `program` and `context`. However, I suppose they are unique in the current iteration. I suppose we can get rid of it.